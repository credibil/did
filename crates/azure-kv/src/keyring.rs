use std::sync::{Arc, Mutex};

use base64ct::{Base64UrlUnpadded, Encoding};
use did_core::error::Err;
use did_core::{tracerr, Jwk, KeyOperation, KeyRing, Result};

use crate::client::KeyVault;
use crate::key_bundle::JsonWebKey;

/// Azure Key Vault key ring.
#[derive(Clone)]
pub struct AzureKeyRing {
    // Azure key vault client.
    pub(crate) client: KeyVault,
    // In-memory storage of key names before commits.
    key_buffer: Arc<Mutex<Vec<String>>>,
    // Namespace for keys in this key ring. This is used to differentiate keys in the same Azure
    // Key Vault instance to support, for example, multi-tenancy.
    namespace: Option<String>,
}

/// Azure Key Vault key ring constructor.
impl AzureKeyRing {
    /// Default constructor.
    ///
    /// # Arguments
    ///
    /// * `client` - Azure Key Vault client. Uses environment variables for credentials.
    /// * `namespace` - If multiple sets of keys can be stored in the same backing store, use a
    /// namespace to differentiate this key ring from adjacent ones. (eg. multi-tenancy)
    #[must_use]
    pub fn new(client: KeyVault, namespace: Option<String>) -> Self {
        Self {
            client,
            key_buffer: Arc::new(Mutex::new(Vec::new())),
            namespace,
        }
    }

    /// Delete a key from the key ring. This is a convenience method for testing.
    ///
    /// # Arguments
    ///
    /// * `op` - The key operation to delete.
    ///
    /// # Errors
    ///
    /// * `Err::RequestError` - A problem occurred calling the Azure Key Vault API.
    /// * `Err::InvalidConfig` - The Azure Key Vault URL or credentials are not configured.
    /// * `Err::AuthError` - A problem occurred authenticating with Azure Key Vault.
    pub async fn delete_key(&self, op: &KeyOperation) -> Result<()> {
        let key_name = self.key_name(op);
        match self.client.delete_key(&key_name).await {
            Ok(_) => Ok(()),
            Err(e) => Err(e),
        }
    }
}

/// Key ring implementation backed by Azure Key Vault.
impl KeyRing for AzureKeyRing {
    /// Get the currently active public key for the specified key operation. If there is no such key
    /// attempt to find the most recent previous version.
    async fn active_key(&self, op: &KeyOperation) -> Result<Jwk> {
        let key_name = self.key_name(op);

        // Get key from Azure Key Vault
        let key = self.client.get_key(&key_name, None).await?;

        // If the key is enabled, return it
        if key.attributes.enabled.is_some_and(|enabled| enabled) {
            return az_to_jwk(key.key);
        }

        // Otherwise the key is disabled, so return the immediate previous version
        let key = self.client.get_previous_version(&key_name).await?;

        az_to_jwk(key.key)
    }

    /// Create or get the next version of the key for the specified operation.
    async fn next_key(&self, op: &KeyOperation) -> Result<Jwk> {
        // Create the key in Azure Key Vault or get the key if it already exists and the latest
        // version is disabled
        self.create_key(op, false).await
    }

    /// Commit (make active) the 'Next' key versions created.
    async fn commit(&self) -> Result<()> {
        // activate each key in the key buffer
        let keys = {
            let kb = self
                .key_buffer
                .lock()
                .expect("lock on key buffer mutex poisoned due to panic in another thread");
            kb.clone()
        };
        for key_name in &keys {
            self.client.activate_key(key_name).await?;
        }
        let mut kb = self
            .key_buffer
            .lock()
            .expect("lock on key buffer mutex poisoned due to panic in another thread");
        kb.clear();
        Ok(())
    }
}

impl AzureKeyRing {
    // Creates a new key with the specified operation and active status. If the key already exists,
    // return the key without changing the active status.
    pub(crate) async fn create_key(&self, op: &KeyOperation, active: bool) -> Result<Jwk> {
        let key_name = self.key_name(op);

        // Check if the key already exists
        match self.client.get_key(&key_name, None).await {
            Ok(key) => az_to_jwk(key.key),
            Err(e) => {
                if e.is(Err::KeyNotFound) {
                    // Create key in Azure Key Vault
                    let key_bundle = self.client.create_key(&key_name, active).await?;

                    // Add the key name to the key buffer
                    let mut kb = self
                        .key_buffer
                        .lock()
                        .expect("lock on key buffer mutex poisoned due to panic in another thread");
                    kb.push(key_name.clone());

                    // Return the key
                    az_to_jwk(key_bundle.key)
                } else {
                    Err(e)
                }
            }
        }
    }

    /// Constructs a conventional name for a key in the key store. Key names should be unique by
    /// operation and (optional) namespace.
    #[must_use]
    pub fn key_name(&self, op: &KeyOperation) -> String {
        match &self.namespace {
            Some(ns) => format!("vc-{op}-{ns}"),
            None => format!("vc-{op}"),
        }
    }
}

pub(crate) fn az_to_jwk(az_key: JsonWebKey) -> Result<Jwk> {
    let crv = if az_key.curve_name.unwrap_or_default().as_str() == "P-256K" {
        "secp256k1"
    } else {
        // LATER: Add support for other curves.
        tracerr!(Err::InvalidKey, "unsupported curve.")
    };
    let x = if let Some(x) = az_key.x {
        Base64UrlUnpadded::encode_string(&x)
    } else {
        tracerr!(Err::InvalidKey, "Missing x coordinate.")
    };
    let y = if let Some(y) = az_key.y {
        Base64UrlUnpadded::encode_string(&y)
    } else {
        tracerr!(Err::InvalidKey, "Missing y coordinate.")
    };
    Ok(Jwk {
        kty: az_key.key_type,
        crv: Some(crv.to_string()),
        x: Some(x),
        y: Some(y),
        ..Default::default()
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_key_ring() -> AzureKeyRing {
        let url = std::env::var("AZURE_KEY_VAULT").expect("AZURE_KEY_VAULT env var not set");
        let namespace = format!("test-{}", uuid::Uuid::new_v4());
        AzureKeyRing::new(KeyVault::new(url.as_str()), Some(namespace))
    }

    #[tokio::test]
    #[ignore]
    async fn active_key() {
        let kr = test_key_ring();

        let op = KeyOperation::Update;
        let key_name = kr.key_name(&op);

        // initially there should be no active key
        match kr.active_key(&op).await {
            Ok(_) => panic!("Expected key not found error from active_key"),
            Err(e) => {
                if !e.is(Err::KeyNotFound) {
                    panic!("Unexpected error from active_key: {}", e);
                }
            }
        };

        // create an active key, add it to the key buffer, and check it is returned without error
        let _new_key = match kr.create_key(&op, true).await {
            Ok(key) => key,
            Err(e) => {
                panic!("Error creating key: {}", e);
            }
        };

        let key_buffer = {
            let kb = kr
                .key_buffer
                .lock()
                .expect("lock on key buffer mutex poisoned due to panic in another thread");
            kb.clone()
        };
        assert!(key_buffer.contains(&key_name));
        let active_key = kr.active_key(&op).await.expect("active key should be Ok");

        // pause to allow Azure Key Vault to create another version with different time stamp since
        // resolution is only to the second
        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;

        // create a next key, add it to the key buffer, and check it is returned without error
        let next_key = kr.next_key(&op).await;
        assert!(next_key.is_ok());

        // calling active_key after next_key should result in the same key version (idempotency)
        let active_key2 = kr.active_key(&op).await.expect("expected active key should be Ok");
        assert_eq!(active_key.x, active_key2.x);
        assert_eq!(active_key.y, active_key2.y);

        // clean up
        for key_name in key_buffer.iter() {
            kr.client
                .delete_key(key_name)
                .await
                .expect("key should have been deleted without error");
        }
    }

    #[tokio::test]
    #[ignore]
    async fn next_key() {
        let kr = test_key_ring();
        let op = KeyOperation::Update;

        // create new key
        let next = kr.next_key(&op).await.expect("failed to get next update key");

        // should be tracking 1 created key
        let key_buffer = {
            let kb = kr.key_buffer.lock().expect("lock on key buffer mutex poisoned");
            kb.clone()
        };
        assert_eq!(key_buffer.len(), 1);

        // key should not be active
        match kr.active_key(&op).await {
            Ok(_) => panic!("Expected key not found error from active_key"),
            Err(e) => {
                if !e.is(Err::KeyNotFound) {
                    panic!("Unexpected error from active_key: {}", e);
                }
            }
        };

        // check key properties
        assert_eq!(next.kty, "EC");
        assert_eq!(next.crv, Some("secp256k1".to_string()));

        // another call to next_key should return the same key
        let next2 = kr.next_key(&op).await.expect("failed to get next expected update key");
        assert_eq!(next.x, next2.x);
        assert_eq!(next.y, next2.y);

        // clean up
        for key_name in key_buffer.iter() {
            kr.client.delete_key(key_name).await.expect("failed to delete key");
        }
    }

    #[tokio::test]
    #[ignore]
    async fn commit() {
        let kr = test_key_ring();

        // create a new key
        let op = KeyOperation::Update;
        let key_name = kr.key_name(&op);
        let next = kr.next_key(&op).await.expect("failed to get next update key");

        // key should not be active
        match kr.active_key(&op).await {
            Ok(_) => panic!("Expected key not found error from active_key"),
            Err(e) => {
                if !e.is(Err::KeyNotFound) {
                    panic!("Unexpected error from active_key: {}", e);
                }
            }
        };

        // expect one key that can be committed
        let key_buffer = {
            let kb = kr.key_buffer.lock().expect("lock on key buffer mutex poisoned");
            kb.clone()
        };
        assert_eq!(key_buffer.len(), 1);

        // commit
        let commit = kr.commit().await;
        assert!(commit.is_ok());

        // expect no keys for committment
        let key_buffer = {
            let kb = kr.key_buffer.lock().expect("lock on key buffer mutex poisoned");
            kb.clone()
        };
        assert_eq!(key_buffer.len(), 0);

        // key should be active
        let active2 = kr.active_key(&op).await.expect("failed to get next update key");

        // next key and currently active key should have the same version
        assert_eq!(next.x, active2.x);
        assert_eq!(next.y, active2.y);

        // clean up
        kr.client.delete_key(&key_name).await.expect("failed to delete key");
    }
}
