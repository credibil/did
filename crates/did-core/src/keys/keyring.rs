//! Key management. This module provides a trait that can be implemented for creating and rolling
//! over cryptographic keys for use in DID operations or general signing and verification.

use crate::keys::{Jwk, KeyOperation};
use crate::Result;

/// Key generation and storage provider. The `self` reference allows for configuration information
/// such as key store location and credentials.
#[allow(async_fn_in_trait)]
pub trait KeyRing {
    /// Get the currently active public key for the specified key operation. If there is no such key
    /// active key in the key ring, attempt to find the next most recent previous version of that
    /// key.
    ///
    /// # Arguments
    ///
    /// * `op` - The key operation type.
    ///
    /// # Returns
    ///
    /// A [`PublicKeyJwk`] containing the public key for the specified key operation or an error if
    /// the key could not be found or some other error occurred.
    async fn active_key(&self, op: &KeyOperation) -> Result<Jwk>;

    /// Get the public key for the next key pair that will be used for DID operations of the
    /// specified type.
    ///
    /// # Arguments
    ///
    /// * `op` - The key operation type.
    ///
    /// # Returns
    ///
    /// A [`PublicKeyJwk`] containing the public key for the specified key operation or an error if
    /// none could be created.
    async fn next_key(&self, op: &KeyOperation) -> Result<Jwk>;

    /// The `KeyRing` is asynchronous and needs to be thread-safe so we cannot mutate the `KeyRing`
    /// structure itself to manage interim key information. Commit will be called following
    /// a successful DID operation and can be used to save newly generated "next" keys to a key
    /// store and make them current. A no-op default is provided if your implementation does not
    /// need a commit.
    async fn commit(&self) -> Result<()> {
        async move { Ok(()) }.await
    }
}
