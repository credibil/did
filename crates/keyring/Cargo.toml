[package]
name = "keyring"
description = "Azure Key Vault implementation of the vercre-did KeyRing and Signer traits"
authors.workspace = true
edition.workspace = true
version.workspace = true

[lints]
workspace = true

[dependencies]
base64ct = "1.6.0"
ecdsa = "0.16.9"
k256 = { version = "0.13.4", features = ["jwk"] }
rand = "0.8.5"
serde_json = "1.0.128"
sha2 = "0.10.8"
tracing.workspace = true
did-core = { version = "0.1.0", path = "../did-core" }

[dev-dependencies]
tokio = { version = "1.40.0", features = ["full"] }
