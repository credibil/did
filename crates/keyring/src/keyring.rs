use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use did_core::error::Err;
use did_core::{Jwk, KeyOperation, KeyRing, Result};

use crate::KeyPair;

/// Ephemeral key ring.
pub struct EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Once a key is generated it can be stored for the scope of the struct.
    pub(crate) current_keys: Arc<Mutex<HashMap<KeyOperation, Arc<K>>>>,
    /// Holds newly generated keys that are not yet active.
    pub(crate) next_keys: Arc<Mutex<HashMap<KeyOperation, Arc<K>>>>,
}

/// Default
impl<K> Default for EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    fn default() -> Self {
        Self::new()
    }
}

/// Configuration and key generation.
impl<K> EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Create a new `EphemeralKeyRing` instance.
    #[must_use]
    pub fn new() -> Self {
        Self {
            current_keys: Arc::new(Mutex::new(HashMap::new())),
            next_keys: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Generate a new key pair.
    fn generate() -> Result<Arc<K>> {
        let kp = K::generate()?;
        Ok(Arc::new(kp))
    }
}

/// `KeyRing` implementation
impl<K> KeyRing for EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Get the current key for the given operation.
    async fn active_key(&self, op: &KeyOperation) -> Result<Jwk> {
        let current_keys = self.current_keys.lock().expect("lock on current_keys mutex failed");
        if current_keys.contains_key(op) {
            return current_keys[op].to_jwk();
        }
        Err(Err::KeyNotFound.into())
    }

    /// Generate a new key pair for the given operation.
    async fn next_key(&self, op: &KeyOperation) -> Result<Jwk> {
        let key = Self::generate()?;
        let mut next_keys = self.next_keys.lock().expect("lock on next_keys mutex failed");
        next_keys.insert(op.clone(), key.clone());
        key.to_jwk()
    }

    /// Commit (make active) the next keys created.
    async fn commit(&self) -> Result<()> {
        let mut current_keys = self.current_keys.lock().expect("lock on current_keys mutex failed");
        let next_keys = self.next_keys.lock().expect("lock on next_keys mutex failed");
        for (op, key) in next_keys.iter() {
            current_keys.insert(op.clone(), key.clone());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::secp256k1::KeyPair as Secp256k1KeyPair;

    #[tokio::test]
    async fn test_keyring() {
        let keyring = EphemeralKeyRing::<Secp256k1KeyPair>::new();
        let first = keyring.active_key(&KeyOperation::Sign).await;
        assert!(first.is_err());
        let next = keyring.next_key(&KeyOperation::Sign).await.unwrap();
        keyring.commit().await.unwrap();
        let active = keyring.active_key(&KeyOperation::Sign).await.unwrap();
        assert_eq!(next, active);
    }
}
