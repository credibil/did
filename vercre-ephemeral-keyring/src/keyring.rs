use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
};
use vercre_didcore::{error::Err, Jwk, KeyOperation, KeyRing, Result};

use crate::KeyPair;

/// Ephemeral key ring.
pub struct EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Once a key is generated it can be stored for the scope of the struct.
    pub(crate) current_keys: Arc<Mutex<HashMap<KeyOperation, Arc<K>>>>,
    /// Holds newly generated keys that are not yet active.
    pub(crate) next_keys: Arc<Mutex<HashMap<KeyOperation, Arc<K>>>>,
}

/// Configuration and key generation.
impl<K> EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Create a new `EphemeralKeyRing` instance.
    #[must_use]
    pub fn new() -> Self {
        Self {
            current_keys: Arc::new(Mutex::new(HashMap::new())),
            next_keys: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Generate a new key pair.
    fn generate(&self) -> Result<Arc<K>> {
        let kp = K::generate()?;
        Ok(Arc::new(kp))
    }
}

/// `KeyRing` implementation
#[allow(async_fn_in_trait)]
impl<K> KeyRing for EphemeralKeyRing<K>
where
    K: KeyPair + Send + Sync,
{
    /// Get the current key for the given operation.
    async fn active_key(&self, op: &KeyOperation) -> Result<Jwk> {
        let current_keys = self.current_keys.lock().expect("lock on current_keys mutex failed");
        if current_keys.contains_key(op) {
            return current_keys[op].to_jwk();
        }
        Err(Err::KeyNotFound.into())
    }

    /// Generate a new key pair for the given operation.
    async fn next_key(&self, op: &KeyOperation) -> Result<Jwk> {
        let key = self.generate()?;
        let mut next_keys = self.next_keys.lock().expect("lock on next_keys mutex failed");
        next_keys.insert(op.clone(), key.clone());
        key.to_jwk()
    }

    /// Commit (make active) the next keys created.
    async fn commit(&self) -> Result<()> {
        let mut current_keys = self.current_keys.lock().expect("lock on current_keys mutex failed");
        let next_keys = self.next_keys.lock().expect("lock on next_keys mutex failed");
        for (op, key) in next_keys.iter() {
            current_keys.insert(op.clone(), key.clone());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::secp256k1::KeyPair as Secp256k1KeyPair;

    #[tokio::test]
    async fn test_keyring() {
        let keyring = EphemeralKeyRing::<Secp256k1KeyPair>::new();
        let first = keyring.active_key(&KeyOperation::Sign).await;
        assert!(first.is_err());
        let next = keyring.next_key(&KeyOperation::Sign).await.unwrap();
        keyring.commit().await.unwrap();
        let active = keyring.active_key(&KeyOperation::Sign).await.unwrap();
        assert_eq!(next, active);
    }
}
