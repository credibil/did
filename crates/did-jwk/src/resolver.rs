use base64ct::{Base64UrlUnpadded, Encoding};
use did_core::{
    DocumentMetadata, Jwk, Resolution, ResolutionMetadata, Resolver, Result, DID_CONTEXT,
};
use keyring::KeyPair;

use crate::{document_from_jwk, Registrar};

/// Resolver implementation for the JWK method.
///
/// # Arguments
///
/// * `did` - The DID to resolve.
///
/// # Returns
///
/// The DID document with a single verification method corresponding to the supplied DID and the
/// corresponding verification method references for the possible purposes implied by the JWK
/// encoded in the DID.
impl<'a, K> Resolver for Registrar<'a, K>
where
    K: KeyPair + Send + Sync,
{
    async fn resolve(&self, did: &str) -> Result<Resolution> {
        if !did.starts_with("did:jwk:") {
            return Ok(error_response("invalidDid"));
        }
        let content_type = "application/did+ld+json".to_string();

        let parts = did.split(':').collect::<Vec<&str>>();
        if parts.len() != 3 {
            return Ok(error_response("invalidDid"));
        }

        let encoded = parts[2];
        let serialized = Base64UrlUnpadded::decode_vec(encoded)?;
        let key: Jwk = serde_json::from_slice(&serialized)?;

        let doc = document_from_jwk(&key, &K::key_type().cryptosuite(), did);
        Ok(Resolution {
            context: DID_CONTEXT.to_string(),
            did_document: Some(doc?),
            did_document_metadata: Some(DocumentMetadata::default()),
            did_resolution_metadata: Some(ResolutionMetadata {
                content_type,
                ..Default::default()
            }),
        })
    }
}

fn error_response(error: &str) -> Resolution {
    Resolution {
        context: DID_CONTEXT.to_string(),
        did_document: None,
        did_document_metadata: Some(DocumentMetadata::default()),
        did_resolution_metadata: Some(ResolutionMetadata {
            content_type: "application/did+ld+json".to_string(),
            error: Some(error.to_string()),
        }),
    }
}

#[cfg(test)]
mod tests {
    use did_core::DID_CONTEXT;
    use keyring::{EphemeralKeyRing, Secp256k1KeyPair};

    use super::*;
    use crate::Registrar;

    #[tokio::test]
    async fn test_resolve_invalid_did_method() {
        let keyring = EphemeralKeyRing::<Secp256k1KeyPair>::new();
        let registrar = Registrar::new(&keyring);
        let resolution = registrar.resolve("did:web:wibble").await.unwrap();
        assert_eq!(resolution.context, DID_CONTEXT);
        assert!(resolution.did_document.is_none());
        assert!(resolution.did_document_metadata.is_some());
        assert!(resolution.did_resolution_metadata.is_some());
        let md = resolution.did_resolution_metadata.unwrap();
        assert_eq!(md.clone().content_type, "application/did+ld+json");
        assert_eq!(md.clone().error.unwrap(), "invalidDid");
    }

    #[tokio::test]
    async fn test_resolve_secp256k1_did() {
        let did = "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ".to_string();
        let keyring = EphemeralKeyRing::<Secp256k1KeyPair>::new();
        let registrar = Registrar::new(&keyring);
        let resolution = registrar.resolve(&did).await.unwrap();
        insta::with_settings!( {sort_maps => true}, {
            insta::assert_yaml_snapshot!(resolution);
        });
    }
}
