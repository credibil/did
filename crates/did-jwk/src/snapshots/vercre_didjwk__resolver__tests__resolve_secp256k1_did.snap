---
source: did-jwk/src/resolver.rs
assertion_line: 93
expression: resolution
---
"@context": "https://www.w3.org/ns/did/v1"
didDocument:
  id: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ"
  "@context": "https://www.w3.org/ns/did/v1"
  verificationMethod:
    - id: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
      type: EcdsaSecp256k1VerificationKey2019
      controller: "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ"
      publicKeyJwk:
        kty: EC
        crv: secp256k1
        x: JJzPi4qy2rvKSVO9F2-05VWeX2ohswX7STo87MGqqPw
        y: C1RtFnqW9lNLB5z72pon123dxv1kDqU3Ql5B8s0Wc_4
  authentication:
    - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
  assertionMethod:
    - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
  keyAgreement:
    - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
  capabilityInvocation:
    - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
  capabilityDelegation:
    - "did:jwk:eyJrdHkiOiJFQyIsImNydiI6InNlY3AyNTZrMSIsIngiOiJKSnpQaTRxeTJydktTVk85RjItMDVWV2VYMm9oc3dYN1NUbzg3TUdxcVB3IiwieSI6IkMxUnRGbnFXOWxOTEI1ejcycG9uMTIzZHh2MWtEcVUzUWw1QjhzMFdjXzQifQ#0"
didDocumentMetadata: {}
didResolutionMetadata:
  contentType: application/did+ld+json
  error: ~

